sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -1029,7 +1040,27 @@ public double getProgress() {
   private Progress getProgress() {
     synchronized (splitLock) {
       if (currentTracker instanceof RestrictionTracker.HasProgress) {
-        return ((HasProgress) currentTracker).getProgress();
+        Progress progress = ((HasProgress) currentTracker).getProgress();
+        double totalWork = progress.getWorkCompleted() + progress.getWorkRemaining();
+        double completed =
+            totalWork * currentWindowIterator.previousIndex() + progress.getWorkCompleted();
+        double remaining =
+            totalWork * (currentElement.getWindows().size() - currentWindowIterator.nextIndex())
+                + progress.getWorkRemaining();
+        return Progress.from(completed, remaining);
+      }
+    }
+    return null;
+  }
+
+  private Progress getProgressFromWindowObservingTruncate(double elementCompleted) {
+    synchronized (splitLock) {
+      if (currentWindowIterator != null) {
```

Comment:
Should we register this with `addProgressRequestCallback` so we generate monitoring infos?

I'm not sure if truncate should be using the downstream progress as part of its calculation when reporting it as a monitoring info. I know that this differs from how we calculate the progress/split point for the SplittableFnDataReceiver since the singular fraction needs to take into account the downstream progress accurately.

I was always envisioning that work completed/work remaining for the monitoring infos always represented the local knowledge of work and didn't take into account any downstream/upstream knowledge. We can avoid this issue if we merge this logic into the `getProgress` method around line 533.

Reply:
I was hesitated to do so since I'm not sure whether the progress from Truncate is used anywhere unless it's used for runner side split decision. Then there is another question, should the progress from Truncate be used for split decision?

Reply:
Thinking about this more, I do believe that the progress does need to be reported as a metric so a runner can choose a split fraction and also compute the amount of remaining work and/or completion time estimate. It looks like either:
1) Need to make work completed/remaining take into account downstream processing
OR
2) Need to add a metric that represents work in progress so that a runner can compute the amount of work being done (without this we can't figure out how much the work remaining downstream is relative to an upstream node).

I'm not a big fan of 1) since it means that this metric is intrinsically tied to the state of other transforms while in 2) we are adding something new.

Reply:
I think the combination of Truncate observing windows and Process consuming elements makes progress/split quite complicated. It seems like a special case of handling split/progress for a non-root SDF.
If we treat the Truncate only as a special case, we can go with option1 and let the runner to decide how to make the split decision. In this case, it's natural for Truncate to report the progress combing with the downstream progress. 

For option2, I'm not sure how a runner computes the work remaining based on the new metric.

Reply:
Yes, truncate exposes some of the issues where a non root SDF makes things interesting.

For option2, if we assume that `inprogress` is included in `remaining` then:
```
total = completed + remaining
fraction_completed = (inprogress / remaining) * downstream_fraction_completed + completed / total
```
`downstream_fraction_completed` would be computed recursively and could also be computed effectively if there were multiple consumers with `(completed_conumers + downstream_fraction_completed) / total_num_consumers`

Reply:
Thanks for the formula. If we let the runner to do the computation, the runner needs to lookup the graph to figure out which progress is from downstream. I'm thinking about another option: we can add a progress signal for split specifically, which should be reported by the root node of the SDK graph. All computations you have mentioned can be done in the SDK side recursively. The runner side can decide whether to look into this signal by knowing whether there is an SDF(or SDFs) in the SDK graph.

Reply:
Originally the idea was that we didn't want the SDK to have to perform these calculations and it is why each operator was going to report work_completed/work_remaining if it had them but it seems like accurate splitting by fraction needs to take it into account.

Using the graph to compute the progress shouldn't be any more/less difficult then the work that is being put into the SDK.

Is there still value in reporting the work_completed/work_remaining metrics then?

Reply:
I did forget something, it turns out that `work_completed`/`work_remaining` are `lists` of progresses so you can report one for each `active` element so it turns out that the protocol didn't forget about this inprogress issue.

Reply:
Are you saying that one transform can report the progress metric per element/window? For example, saying that the Truncate is on window1 and there are 3 windows in total, the downstream sdf progress is 0.5. Then the Truncate should report [(work_completed=1, work_remaining=0), (work_completed=1, work_remaining=0), (work_completed=0, work_remaining=1)] and the sdf should report [(work_completed=0.5, work_remaining=0.5)]? The the actual completed should be 1 + 1 * 0.5?

Reply:
Yes, assuming you meant to say that truncate is on window2.

Reply:
I would prefer to let gRPC report element progress since it looks simpler, and also helps us to report less metrics. It's also more straight forward for each transform to compute element progress regarding to downstream progress.